File: src/main/java/me/andre111/d20server/command/RollCommand.java
package me.andre111.d20server.command;

import me.andre111.d20common.model.chat.ChatEntry;
import me.andre111.d20common.model.profile.Profile;
import me.andre111.d20common.scripting.Context;
import me.andre111.d20common.scripting.expression.Expression;
import me.andre111.d20common.scripting.expression.Parser;
import me.andre111.d20common.scripting.expression.Result;
import me.andre111.d20server.service.ChatService;
import me.andre111.d20server.util.RollFormatter;

public class RollCommand extends Command {
	private final boolean showPublic;
	private final boolean showSelf;
	private final boolean showGM;
	
	private final Parser parser;
	
	public RollCommand(String name, String[] aliases, boolean showPublic, boolean showSelf, boolean showGM) {
		super(name, aliases);
		
		this.showPublic = showPublic;
		this.showSelf = showSelf;
		this.showGM = showGM;
		
		this.parser = new Parser();
	}

	
	@Override
	public void execute(Profile profile, String arguments) {
		// parse roll and execute
		Result result = null;
		Exception exception = null;
		try {
			Expression expr = parser.parse(arguments);
			result = expr.eval(new Context(profile, profile.getMap(), null));
		} catch(Exception e) {
			exception = e;
		}
		
		String rollMessage = RollFormatter.formatDiceRoll(profile, arguments, showPublic, result, exception);
		
		// determine recipents
		long[] recipents = buildRecipents(profile, showPublic, showSelf);
		
		// create entry
		ChatEntry entry = new ChatEntry(rollMessage, profile.id(), showGM, recipents);
		if(result != null) entry.setRolls(result.getDiceRolls());
		
		// append message
		ChatService.append(true, entry);
	}
}

File: core/server/command/roll-command.js
// @ts-check
import { Command } from './command.js';
import { RollFormatter } from '../util/roll-formatter.js';
import { ChatService } from '../service/chat-service.js';

import { ChatEntry } from '../../common/message/chat/chat-entry.js';
import { Scripting } from '../../common/scripting/scripting.js';
import { I18N } from '../../common/util/i18n.js';

const SCRIPT = new Scripting();
export class RollCommand extends Command {
    #showPublic;
    #showSelf;
    #showGM;
    #sendMessage;

    constructor(name, aliases, showPublic, showSelf, showGM) {
        super(name, aliases, false);

        this.#showPublic = showPublic;
        this.#showSelf = showSelf;
        this.#showGM = showGM;
        this.#sendMessage = showPublic || showSelf || showGM;
    }

    execute(profile, args) {
        // parse roll and execute
        const result = SCRIPT.interpretExpression(ChatService.unescape(args), profile, null);
        const diceRolls = SCRIPT.diceRolls;
        if (SCRIPT.errors.length != 0) {
            var error = SCRIPT.errors[0];
            if (SCRIPT.errors.length > 1) {
                error = error + `\nand ${SCRIPT.errors.length - 1} more`
            }
            ChatService.appendError(profile, error);
            return;
        }

        if (this.#sendMessage) {
            const rollAppendix = this.#showPublic ? '' : (this.#showGM ? I18N.get('command.roll.to.gm', ' (to GM)') : I18N.get('command.roll.to.self', ' (to Self)'));
            const rollMessage = RollFormatter.formatDiceRoll(profile, args, result, rollAppendix);

            // determine recipents
            const recipents = this.buildRecipents(profile, this.#showPublic, this.#showSelf);

            // create entry
            const entry = new ChatEntry(rollMessage, profile.getID(), this.#showGM, recipents);
            entry.setRolls(diceRolls);

            // append message
            ChatService.append(true, entry);
        }
    }
}
